File: backend/api/app/domain/use_cases/user/followers.py
from typing import Iterable
from core.use_cases.use_case import UseCase
from app.domain.repositories.user import BaseRepository
from core.common.equatable import Equatable
from core.common.either import Either
from core.errors.failure import Failure
from app.domain.entities.user import User, UserEntity

class Params(Equatable):
    def __init__(self, user_id: str) -> None:
        self.user_id = user_id

class UserFollowers(UseCase[Iterable[UserEntity]]):
    def __init__(self, repository: BaseRepository):
        self.repository = repository
    
    async def __call__(self, params: Params) -> Either[Failure, Iterable[UserEntity]]:
        return await self.repository.followers(params.user_id)

File: mobile/lib/features/architect/domains/use_cases/user/followers.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';

import '../../../../../core/errors/failure.dart';
import '../../../../../core/use_cases/usecase.dart';
import '../../entities/user.dart';
import '../../repositories/user.dart';

class UserFollowers implements UseCase<List<User>, Params> {
  final UserRepository repository;

  const UserFollowers(this.repository);

  @override
  Future<Either<Failure, List<User>>> call(Params params) async {
    return await repository.followers(params.id);
  }
}

class Params extends Equatable {
  final String id;

  const Params(this.id);

  @override
  List<Object?> get props => [id];
}
